grug/jellyfish-merkle/spec/invariants.md

```diff
@@ -0,0 +1,583 @@
+
+This also means that the three is dense, not sparse.
+
+*Status:* TRUE
+
+```bluespec apply_state_machine.qnt +=
+  /// If a node has exactly one child, the child is an internal node
+  /// (if it was a leaf, then the node itself would be the leaf)
+  val densityInv: bool = {
+    pure def isDense(t: TreeMap): bool =
+      t.keys().forall(nId => {
+        match t.get(nId) {
+          | Internal(n) =>
+            if (n.left_child == None and n.right_child != None) {
+              // Only has right child, right child should be internal
+              findNode(t, nId.key_hash.append(1)).isInternal()
+            } else if (n.right_child == None and n.left_child != None) {
+              // Only has left child, left child should be internal
+              findNode(t, nId.key_hash.append(0)).isInternal()
+            } else {
+              // Has two children or none
+              true
+            }
+          | Leaf(_) => true
+        }
+      })
+
+    activeVersionedTrees.forall(isDense)
+  }
+```
+<!--
+This inserts a line break that is not rendered in the markdown
+```bluespec apply_state_machine.qnt +=
+
+```
+-->
+
+### Versions of predecessors in path should be >= node version
+
+When nodes are updated (inserted, deleted) they are given a new version, and so are all the predecessors in the tree. At the same time, subtrees that are not touched by an update maintain their version. As a result, if the tree is properly maintained, the parent of a node, should have a version that is greater than or equal to the version of the node.
+
+*Status:* FALSE due to https://github.com/left-curve/left-curve/pull/291
+
+```bluespec apply_state_machine.qnt +=
+  /// Invariant: For every node has predecessors with higer (or equal) version
+  /// (This could be rewritten to talk about direct predecessors only)
+  val versionInv: bool = {
+    pure def allPrefixes (l: List[a]): Set[List[a]] =
+      0.to(l.length()).map(i => l.slice(0, i))
+
+    // This invariant actually works on the whole tree rather than on a TreeMap
+    tree.nodes.keys().forall(a =>
+      allPrefixes(a.key_hash).forall(p =>
+        tree.nodes.keys().exists(b =>
+          p == b.key_hash and b.version >= a.version)))
+  }
+```
+<!--
+This inserts a line break that is not rendered in the markdown
+```bluespec apply_state_machine.qnt +=
+
+```
+-->
+
+### Internal nodes share the version with at least one child
+
+Given the explanation around `versionInv` from above, we had the (wrong) intuition, that since updates always push their version up the tree, every internal node should have the version of at least one of it children. However, the intuition is misleading in the case of a delete, where a parent gets a new version, but there may not be a node at the spot where the deleted nodes had been. However, for reference, we keep the formula here as it might be useful for understanding in the future.
```

Comment:
```suggestion
Given the explanation around `versionInv` from above, we had the (wrong) intuition, that since updates always push their version up the tree, every internal node should have the version of at least one of it children. However, the intuition is misleading 
- in the case of a delete, where a parent gets a new version, but there may not be a node at the spot where
    the deleted nodes had been. However, for reference, we keep the formula here as it might be useful for 
    understanding in the future.
- in the case of adding an empty batch, where we get a new root node at the new version, but the root 
    node's the two subtrees are unchanged.
```

Reply:
Updating your version to move the `However...` part below
```suggestion
Given the explanation around `versionInv` from above, we had the (wrong) intuition, that since updates always push their version up the tree, every internal node should have the version of at least one of it children. However, the intuition is misleading 
- in the case of a delete, where a parent gets a new version, but there may not be a node at the spot where the deleted nodes had been. 
- in the case of applying an empty batch, where we get a new root node at the new version, but the root node's subtrees are unchanged.
However, for reference, we keep the formula here as it might be useful for understanding in the future.
```